Authenticated WebView improvement [Part 2]

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #13467 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR builds on top of the changes added in #13468 to implement the improvements suggested in pe5sF9-3QB-p2, the improvements allow us to handle auto authentication in the WebView in the following cases:
- WordPress.com authenticated:
    - Atomic websites.
    - wordpress.com supported URLs.
    - Jetpack SSO sites.
- Site Credentials authenticated (assuming native login worked for the authentication) 

### Steps to reproduce
> [!IMPORTANT]
> - To avoid any false positives, please ensure that you are signed out from the WebView after each test: Either by clearing the app data, or by tapping the WordPress logout button manually in the WebView.
> - If you changed the Jetpack SSO module settings for a site, ensure that you restart the app to fetch the new site modules.
> - WPS Hide Login that I used in some tests below has issues with redirection, so expect some issues. I couldn't find an alternative plugin that works better, in the past we used to use `WP Hide & Security Enhancer ` which worked well, but its free version doesn't seem to work with Jurassic Ninja sites nor Pressable sites anymore.
> - Jetpack SSO settings can be accessed from Jetpack Settings in wp-admin (`wp-admin/admin.php?page=jetpack#/settings`)
 
##### TC1: WordPress.com authenticated with a wordpress.com supported link
1. Sign in using WordPress.com
2. Open the payments hub.
3. Tap on "order card reader"
4. Tap on Log In in the hamburger menu.
5. Confirm that your WordPress.com account is signed in and suggested as first option.

##### TC2: WordPress.com authenticated with an atomic website
1. Sign in using WordPress.com
2. Select an atomic website.
3. Open the dashboard screen.
4. If needed, add the performance card.
5. Tap on "View all store analytics"
6. Tap on "See report" in one of the analytics cards.
7. Confirm that you are auto authenticated to your account.

##### TC3: WordPress.com authenticated with a Jetpack SSO site
1. Sign in using WordPress.com
2. Select a Jetpack website where Jetpack SSO is enabled (can be enabled from Jetpack settings in wp-admin).
3. Open the dashboard screen.
4. If needed, add the performance card.
5. Tap on "View all store analytics"
6. Tap on "See report" in one of the analytics cards.
7. Confirm that you are auto-authenticated to your account.

##### TC4: WordPress.com authenticated with a Jetpack site without SSO
1. Sign in using WordPress.com
2. Select a Jetpack website where Jetpack SSO is disabled (can be disabled from Jetpack settings in wp-admin).
3. Open the dashboard screen.
4. If needed, add the performance card.
5. Tap on "View all store analytics"
6. Tap on "See report" in one of the analytics cards.
7. Confirm that the page is opened in a Chrome Custom tab.

##### TC5: Site Credentials authenticated using web authorization
1. Use a self-hosted site.
2. Install `WPS Hide Login` plugin.
3. Open WPS Hide Login settings, and use a custom login URL **(Enter the same URL in the redirection URL too)**.
4. Open the app and sign in (you might need to perform the Web Authorization twice, because WPS Hide Login doesn't work well with the redirection, but I didn't find a better alternative)
5. Open the dashboard screen.
6. If needed, add the performance card.
7. Tap on "View all store analytics"
8. Tap on "See report" in one of the analytics cards.
9. Confirm that the page is opened in a Chrome Custom tab.

##### TC6: Site Credentials authenticated natively
1. Use a self-hosted site that doesn't use any security measures that could break the native login.
2. Sign in using site credentials.
3. Open the dashboard screen.
4. If needed, add the performance card.
6. Tap on "View all store analytics"
7. Tap on "See report" in one of the analytics cards.
8. Confirm that you are auto-authenticated to your account.

##### TC7: Site Credentials authenticated natively, but uses custom login URL now
1. Use a self-hosted site that doesn't use any security measures that could break the native login.
2. Sign in using site credentials.
3. Install `WPS Hide Login` plugin.
4. Configure WPS Hide Login to use a custom login URL **(Enter the same URL in the redirection URL too)**.
5. Open the dashboard screen.
6. If needed, add the performance card.
7. Tap on "View all store analytics"
8. Tap on "See report" in one of the analytics cards.
9. Confirm that the login screen is shown (which is expected in this case given the custom login URL).

### Testing information
Check above.

### The tests that have been performed
Check above.

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
First pass at the logic to decide which authentication algorithm to use

hichamboushaba committed:
Handle Jetpack SSO login

hichamboushaba committed:
Handle site credentials authentication

hichamboushaba committed:
Update SiteModel to include the list of Jetpack Modules

hichamboushaba committed:
Add implementation of the supportsJetpackSSO function

hichamboushaba committed:
Handle additional case where the site credentials login fails

If for example the site was updated later with a custom login page

hichamboushaba committed:
Add a utility class to check if auto-authentication is supported

hichamboushaba committed:
Make sure not to save the user's password when using web authorization

When using web authorization, we can't know if the provided password is a correct one or not.

hichamboushaba committed:
Check if the URL is part of the site before accepting authentication

hichamboushaba committed:
Introduce an extension to url encode Strings

hichamboushaba committed:
Add unit tests

hichamboushaba committed:
Simplify logic

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/webview/CanAutoAuthenticateInWebView.kt:
```diff
@@ -0,0 +1,15 @@
+/**
+ * A utility use-case to allow consumers to know beforehand if a URL can be auto-authenticated in a WebView.
+ */
+class CanAutoAuthenticateInWebView @Inject constructor(
```
This usecase makes it easier to check if auto authentication can be used or not, for now, this is used just in the Analytics Hub screen, but I intend to add more usages in future PRs where it makes sense, starting with Product Previews.

hichamboushaba committed:
Fix detekt issues

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt:
```diff
@@ -210,11 +210,7 @@ class LoginSiteCredentialsViewModel @Inject constructor(
-                            password = state.password,
-                            detectedErrorMessage = authenticationError?.errorMessage
-                        )
+                        fetchSiteForTutorial(detectedErrorMessage = authenticationError?.errorMessage)
```
When the application passwords tutorial were introduced, we started to wrongly save the username/password of the users even when using Web Authorization, which shouldn't happen, as we don't know if the credentials are correct, and even if they are correct, they can't be used by the app. This change fixes this issue, and it's needed for a better handling of these sites during the authentication: to know whether we can authenticate the user or not.

hichamboushaba committed:
Fix SiteRestClient unit tests

hichamboushaba committed:
Fix unit tests

irfano reviewed (commented):
- While testing Step **TC5**, I tried to connect Jetpack but encountered a crash. On `trunk`, it doesn't crash, but the JP connection fails. Can you check if this crash is related to this PR?

<details>
  <summary>Crash video</summary>
  
[Screen_recording_20250210_165518.webm](https://github.com/user-attachments/assets/65815b4b-7b4d-4dad-a7a1-24a995ce8004)
  
</details>

```
FATAL EXCEPTION: main (Ask Gemini)
Process: com.woocommerce.android, PID: 7069
java.lang.IllegalArgumentException: Site not cached
	at com.woocommerce.android.ui.login.jetpack.main.JetpackActivationMainViewModel$site$1.invokeSuspend(JetpackActivationMainViewModel.kt:96)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at android.os.Handler.handleCallback(Handler.java:959)
	at android.os.Handler.dispatchMessage(Handler.java:100)
	at android.os.Looper.loopOnce(Looper.java:232)
	at android.os.Looper.loop(Looper.java:317)
	at android.app.ActivityThread.main(ActivityThread.java:8705)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:580)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:886)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@e87ba6a, Dispatchers.Main.immediate]
``` 

- For testing **TC7**, I continued from **TC6**, but at the end, I saw the logged-in state instead of the login screen. Did I do something wrong?

hichamboushaba committed:
Add ability to keep the behavior of catching all exceptions in tests

This is the Coroutines Test default behavior, and it's useful for some tests.

hichamboushaba committed:
Fix issue with Jetpack Connection when site doesn't have username/pwd

hichamboushaba committed:
Improve and fix test

hichamboushaba commented:
Thanks @irfano for the thorough test 🙏 

> While testing Step TC5, I tried to connect Jetpack but encountered a crash. On trunk, it doesn't crash, but the JP connection fails. Can you check if this crash is related to this PR?

Great catch, this was an issue caused by my fix to the Site Credentials login when using the Web Authorization, in the Jetpack action we had a logic that checks that the credentials are present, it's useful for when using the flow from the Site Picker, but not so when connecting using Application Passwords, I pushed the commit 43bbcd9da60d3c01fb574c989a5dcf77a38c9064 to fix the issue and added tests, the tests required some other changes (check them please, and let me know if you have any questions or remarks).

> For testing TC7, I continued from TC6, but at the end, I saw the logged-in state instead of the login screen. Did I do something wrong?

You didn't do anything wrong, it was an issue with the steps, I updated TC7, we can't continue from TC6, as the WebView will be already authenticated, so we need to clear the app data then sign in (or alternatively sign out from the WebView before).

irfano commented:
I tested **TC5** again. It is no longer crashing but still fails to redirect to the Jetpack connection after logging in. This behavior is the same as in `trunk` and is expected.

When testing **TC5** and **TC7**, after logging into the site, the “See report” button opened the Custom Chrome tab as you mentioned. But it was not auto-authenticated. I entered my credentials once, and on subsequent attempts, tapping the “See report” button resulted in auto-authentication. I assume this is expected.

I’ll continue reviewing the code.

hichamboushaba commented:
> I tested TC5 again. It is no longer crashing but still fails to redirect to the Jetpack connection after logging in. This behavior is the same as in trunk and is expected.

Yes, this is expected, the Jetpack Connection URL we use for this case has many issues, and fails many times, it's probably something we need to investigate.

> When testing TC5 and TC7, after logging into the site, the “See report” button opened the Custom Chrome tab as you mentioned. But it was not auto-authenticated. I entered my credentials once, and on subsequent attempts, tapping the “See report” button resulted in auto-authentication. I assume this is expected.

Yes, that's expected, the Chrome Custom tab runs in Chrome, so if you weren't authenticated before there, then you would need to authenticate.

Thanks for the review @irfano

hichamboushaba committed:
Revert "Add ability to keep the behavior of catching all exceptions in tests"

This reverts commit 44c8d81cb14880aca9c53fbc3505abc92336ec30.

hichamboushaba committed:
Use an alternative way for the tests

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/webview/WebViewAuthenticationFlowResolver.kt:
```diff
@@ -0,0 +1,52 @@
+        get() = listOf("wordpress.com", "wp.com", "jetpack.com", "woocommerce.com")
+
+    fun resolve(url: String): WebViewAuthenticationFlow {
+        val currentSite = selectedSite.getOrNull()
```
**Optional suggestion:** I suggest using `selectedSite.getOrNull() ?: return WebViewAuthenticationFlow.None` here, as it would help manage the nullability of `currentSite` below. I mention this because I noticed warnings further down. I guess you added `contract` for this but it seems the IDE doesn't work well with `contract`. 
![image](https://github.com/user-attachments/assets/8d2cf081-1aaa-4ea3-bc3b-b07f3d00f3c3)

hichamboushaba replied:
`selectedSite.getOrNull() ?: return WebViewAuthenticationFlow.None` won't work well, because we allow signing in to WordPress.com URLs even when the user didn't select a site yet, which can happen during the Jetpack Connection.

> I guess you added contract for this but it seems the IDE doesn't work well with contract.

Yes, I added it, I think contracts work well with the IDE (as the official SDK functions depend on them greatly), maybe you just need to clean the build or the IDE cache?

irfano replied:
> selectedSite.getOrNull() ?: return WebViewAuthenticationFlow.None won't work well

👍🏻 

> maybe you just need to clean the build or the IDE cache?

I did both, but still see errors. I'm using "Android Studio Ladybug Feature Drop | 2024.2.2". It's strange that you don't see these errors. What's your AS version?

hichamboushaba replied:
I'm using Android Studio Meerkat | 2024.3.1 Beta 1, but I don't think this will have an impact on this.

Can you check if you enabled the K2 mode in the Kotlin settings of the IDE? I tested by disabling it on my side, and it caused a similar issue to what you are reporting.
K2 mode is now used by the app, so to have a consistent experience, it should be enabled also on the IDE level.

-----
And on a side note, is this a blocking point? 😅 the PR is not approved yet, and there two others PRs that depend on these changes, so I just want to confirm if delaying the approval is intended or not 🙂.

irfano replied:
Good catch, that was the reason, the warning disappeared when I enabled K2 mode. 
I had concerns about opting into  `ExperimentalContracts` and using K2 mode which is still a beta feature. That’s why I didn’t approve it initially. But it looks like you’ve been using K2 mode without any issues for a while, so it seems fine. I’m approving it now. Thanks for the discussion!

hichamboushaba replied:
Thanks @irfano 🙏 

> I had concerns about opting into ExperimentalContracts

Even though the contracts is experimental, it's been here since Kotlin 1.3, and it's used a lot by the Kotlin SDK libraries ([commit](https://github.com/JetBrains/kotlin/commit/fb03656e997fb79308673401278947c92cb78a73) that used them the standard library and [commit](https://github.com/JetBrains/kotlin/commit/accbd07b2e8cbeed670e261a0a34e2aaacbe7d06) that exposed it for developers), so I wouldn't worry about the fact that it says experimental.

> using K2 mode which is still a beta feature

K2 is stable now, and it has been for a while, and by using Kotlin 2.0, we are using it by default.
The IDE support was late to the party, and since you are still using `2024.2` version, it still says experimental, but it [will be](https://blog.jetbrains.com/idea/2024/11/k2-mode-becomes-stable/) stable starting `2024.3` (Meerkat).

Now for the issue itself, the contracts don't depend on K2, I think what's happening here is a problem of consistency: Kotlin 2.0 will use K2, but the IDE wasn't using it, which caused the issue. But I can't prove this, it's just a thoery.

irfano replied:
Thanks for the additional information and references, @hichamboushaba! Everything makes sense!

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/webview/WebViewAuthenticator.kt:
```diff
@@ -53,7 +55,26 @@ class WebViewAuthenticator @Inject constructor(
+
+        // Handle SSO login and redirect back to the original URL
+        val site = selectedSite.get()
+        webViewCookieManager.setCookie(site.url, "jetpack_sso_redirect_to=$url")
```
This `setCookie()` function is persistent, right? Should we remove this cookie if the user disables SSO?

hichamboushaba replied:
This cookie is [removed](https://github.com/Automattic/jetpack/blob/ccaf998817efc80eda90534d5a420a799d6ffa17/projects/packages/connection/src/sso/class-sso.php#L690) by Jetpack itself after handling the login, so we shouldn't care about it.

irfano commented on libs/fluxc/src/main/java/org/wordpress/android/fluxc/model/SiteModel.java:
```diff
@@ -897,6 +900,15 @@ public void setActiveJetpackConnectionPlugins(String activeJetpackConnectionPlug
         mActiveJetpackConnectionPlugins = activeJetpackConnectionPlugins;
     }
 
+    @Nullable
+    public String getJetpackModules() {
+        return mJetpackModules;
+    }
```
I found an issue with SSO. When I disable SSO from Jetpack, `jetpackModules` are not refreshed in the app, even after relaunching it. However, when I switch to a different site and then select the original site, `jetpackModules` are updated. This test case passed because the webview was auto-authenticated after disabling SSO and reselecting the site.
Should we open another issue to refresh `jetpackModules`?

hichamboushaba replied:
This is a bigger issue than this task, and it's not only about `jetpack_modules`, it's about everything related to the site, as currently on app launch, we'll load the site into memory in `SelectedSite`, and we don't update it after the [call](https://github.com/woocommerce/woocommerce-android/blob/81619ae647fad467d9846d4e3c2b8bc5429d4af2/WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt#L174) to refresh the sites is done.
I first thought about adding this logic in this PR, but then decided it's better to handle it separately, as it needs proper testing, I should've mentioned this before, sorry.

I'll create a ticket to keep track of this.

hichamboushaba replied:
I created this [ticket](https://github.com/woocommerce/woocommerce-android/issues/13523) to keep track of this issue.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/common/webview/WebViewAuthenticator.kt:
```diff
@@ -95,6 +99,15 @@ class WebViewAuthenticator @Inject constructor(
+
+        val event = webViewEvents.first { it is WCWebViewEvent.PageFinished || it is WCWebViewEvent.UrlFailed }
+        if (event is WCWebViewEvent.UrlFailed && event.url == site.loginUrlOrDefault) {
+            // In case we failed to authenticate, load the original URL
+            // The failure could happen if  some other security measures were added that would prevent
+            // native handling of login (like using a custom login page or a captcha)
```
Thanks for adding this useful comment. 👍🏻

irfano reviewed (commented):
Great work, @hichamboushaba! The flow in the code is well organized and easy to understand, even though it’s complex. All test cases have passed. I’ve added some minor suggestions and questions.

hichamboushaba committed:
Merge branch 'trunk' into issue/13467-improve-authenticated-webview-2

hichamboushaba committed:
Add a release note
